SimpleGame/src/main/scala-2.11/edu/gemini/experimental/jluhrs/Game.scala

```diff
@@ -0,0 +1,92 @@
+package edu.gemini.experimental.jluhrs
+
+import EventStream._
+import Timer._
+import scalaz._
+import Scalaz._
+
+/**
+  * Created by jluhrs on 5/24/16.
+  */
+
+/* The game logic is here.
+ * It's just an object that moves at constant speed, and refresh its position at given times.
+ * The direction of movement is affected by some input events.
+ */
+object Game {
+
+  type Position = (Double, Double)
+  type SpeedDir = (Double, Double)
+
+  case class GameState(t: Time, position: Position, speedDir: SpeedDir, velocity: Double, boardWidth: Int, boardHeight: Int)
+
+  sealed trait PlayerInput
+
+  sealed trait DirectionalInput extends PlayerInput
+
+  case object Up extends DirectionalInput
+  case object Down extends DirectionalInput
+  case object Right extends DirectionalInput
+  case object Left extends DirectionalInput
+
+  implicit def decodeSpeedDir(i: DirectionalInput): SpeedDir = i match {
```

Comment:
This is to allow you to use a `DirectionalInput` where a `SpeedDir` is expected?  This kind of implicit conversion is generally discouraged because it can be hard to follow.  You can make the transition explicit by adding a method `toSpeedDir` in `DirectionalInput` like so:

````scala
sealed trait DirectionalInput extends PlayerInput {
  def toSpeedDir: SpeedDir = this match {
    // conversion logic here
  }
}
````

or I guess just remove `implicit` and call this directly when you need it.